test/functional/wallet_coinselection.py

```diff
@@ -0,0 +1,261 @@
+        # so it's still efficient to drop that much more to fees
+        tx = select_coins_for(0.99996201)
+        # TODO: implement optimization and uncomment the assertions
+        # assert_equal(len(tx['vin']), 1)
+        # assert_equal(len(tx['vout']), 1)
+        # assert(tx['vin'][0]['txid'] in [tx1, tx2])
+
+        # fees for 1 legacy input tx = (148+42)*20 = 3800sat
+        # at this point we're dropping to fees enough so we can spend heavier input with less waste
+        tx = select_coins_for(0.99996200)
+        assert_equal(len(tx['vin']), 1)
+        assert_equal(len(tx['vout']), 1)
+        assert_equal(tx['vin'][0]['txid'], tx3)
+
+        # paying enough fees for 2x bech32 input tx
+        # but using one legacy and dropping for fees is still more efficient in terms of waste
+        tx = select_coins_for(0.99995820)
+        assert_equal(len(tx['vin']), 1)
+        assert_equal(len(tx['vout']), 1)
+        assert_equal(tx['vin'][0]['txid'], tx3)
+
+        # max amount when dropping for fees is still more efficient than creating a change
+        # With fee = 4980, we'd be dropping 1180sat excess which brings waste to 2660
+        # which is equal to the waste of 2x bech32 solution
+        tx = select_coins_for(0.99995021)
+        assert_equal(len(tx['vin']), 1)
+        assert_equal(len(tx['vout']), 1)
+        assert_equal(tx['vin'][0]['txid'], tx3)
+
+        # a solution with two bech32 inputs and a change is the most efficient
+        # but it could only be found by SRD
+        tx = select_coins_for(0.99995020)
+        # TODO: randomness here, could be one or two inputs
+        # print(tx['vin'])
+        # print("{} / {}".format(len(tx['vin']), len(tx['vout'])))
+
+        # With fee at 5101sat we would be dropping 1301sat for fess
+        # This is more than cost of change and BnB won't find a solution with single input
+        # We either get
+        # 1) 2x bech32 solution (SRD only)
+        # 2) legacy + bech32 solution
+        tx = select_coins_for(0.99994899)
+        # legacy + dropping more for fees is still better in terms of waste than legacy + bech32
+        # TODO: implement optimization and update asserts
+        assert_equal(len(tx['vin']), 2)
+        assert_equal(len(tx['vout']), 2)
+
+        # a solution with change is the best solution
+        # 2x bech32 is still more efficient than bech32-and-legacy, but only could be found with SRD
+        tx = select_coins_for(0.99994220)
+        assert_equal(len(tx['vin']), 2)
+        assert_equal(len(tx['vout']), 2)
+
+    def test_one_big_and_many_small_coins(self):
+        w0 = self.nodes[0].get_wallet_rpc(self.default_wallet_name)
+        self.nodes[0].createwallet(wallet_name="test_one_big_and_many_small_coins")
+        w1 = self.nodes[0].get_wallet_rpc("test_one_big_and_many_small_coins")
+
+        # Create a bunch of small coins
+        for i in range(100):
```

Comment:
Actually I was trying to keep the number of small coins as a low as possible while still demonstrating some inefficiencies of coin selection.
1) The more coins we have -> the more time the test takes.
2) The more coins we have -> the harder it's to debug and to reason about.

What is the test scenario that would be possible with 1000 coins, but not possible with 100 coins?

P.S. I think we can definitely reproduce this scenario, I see no reasons why not? If utxo pool is equivalent to this test cases then even with SRD it will find better solution in ~856/1000 = 85%, which leaves 15% to a really bad knapsack alternative.

Reply:
Yeah, I think it would be easy to produce the scenario: 

• 100 small coins, 1 big coin
• no BnB solution
• a large set of the small coins produces a smaller change output than the 1 big coin, so knapsack prefers the former
• SRD only finds the 1 big coin by itself as a solution in ~1/101 cases

⇒ usually a worse waste than best solution, but better than knapsack

Reply:
Isn't this exactly #3 in `test_one_big_and_many_small_coins`?